File: packages/sb-uhq/src/self/commands/utilitaires/tag.js
"use strict";

const fs   = require("fs");
const path = require("path");
const { dataPath } = require("../../func/data-path");
const { messageEdit } = require("../../func/message_edit");

const TAGS_FILE = dataPath("config", "tags.json");

// ── I/O ──────────────────────────────────────────────────────────────────────

function load() {
  try { return JSON.parse(fs.readFileSync(TAGS_FILE, "utf8")); }
  catch { return {}; }
}

function save(tags) {
  fs.mkdirSync(path.dirname(TAGS_FILE), { recursive: true });
  fs.writeFileSync(TAGS_FILE, JSON.stringify(tags, null, 2));
}

// ── Logique pure ─────────────────────────────────────────────────────────────

/**
 * @param {import("discord.js-selfbot-v13").Client} client
 * @param {{ action: "add"|"remove"|"edit"|"list"|"send", name?: string, content?: string, channelId?: string }} payload
 */
async function execute(client, payload) {
  const { action, name, content, channelId } = payload;
  const tags = load();

  if (action === "list") return { tags };

  if (action === "add") {
    if (!name || !content) throw new Error("name et content requis.");
    if (tags[name]) throw new Error(`Le tag '${name}' existe déjà.`);
    tags[name] = content;
    save(tags);
    return { tags };
  }

  if (action === "edit") {
    if (!name || !content) throw new Error("name et content requis.");
    if (!tags[name]) throw new Error(`Tag '${name}' introuvable.`);
    tags[name] = content;
    save(tags);
    return { tags };
  }

  if (action === "remove") {
    if (!name) throw new Error("name requis.");
    if (!tags[name]) throw new Error(`Tag '${name}' introuvable.`);
    delete tags[name];
    save(tags);
    return { tags };
  }

  if (action === "send") {
    if (!name) throw new Error("name requis.");
    if (!tags[name]) throw new Error(`Tag '${name}' introuvable.`);
    if (!channelId) throw new Error("channelId requis pour envoyer un tag.");
    const channel = await client.channels.fetch(channelId).catch(() => null);
    if (!channel) throw new Error(`Salon ${channelId} introuvable.`);
    await channel.send(tags[name]);
    return { sent: true, name, channelId };
  }

  throw new Error(`Action tag inconnue : '${action}'`);
}

// ── Handler commande préfixe ──────────────────────────────────────────────────

async function callback(client, message, args) {
  if (!args.length) return messageEdit(message, "`❌` **Usage :** `.tag <nom>`");

  const [name] = args;
  const tags   = load();

  if (!tags[name]) return messageEdit(message, `\`❌\` **Tag \`${name}\` introuvable.**`);

  return message.edit(tags[name]).catch(() => message.channel.send(tags[name]));
}

module.exports = { name: "tag", execute, callback };

File: packages/sb-uhq/app/commands/utilitaires/tag.py
"""tag — messages prédéfinis (bridge + commande préfixe)."""

from __future__ import annotations

from ...func.data_path import data_path, read_json, write_json
from ...func.discord_util import fetch_channel
from ...func.message_edit import message_edit

TAGS_FILE = data_path("config", "tags.json")


def _load() -> dict:
    return read_json(TAGS_FILE, {})


def _save(tags) -> None:
    write_json(TAGS_FILE, tags)


async def execute(client, payload):
    action = payload.get("action")
    name = payload.get("name")
    content = payload.get("content")
    channel_id = payload.get("channelId")
    tags = _load()

    if action == "list":
        return {"tags": tags}

    if action == "add":
        if not name or not content:
            raise ValueError("name et content requis.")
        if name in tags:
            raise ValueError(f"Le tag '{name}' existe déjà.")
        tags[name] = content
        _save(tags)
        return {"tags": tags}

    if action == "edit":
        if not name or not content:
            raise ValueError("name et content requis.")
        if name not in tags:
            raise ValueError(f"Tag '{name}' introuvable.")
        tags[name] = content
        _save(tags)
        return {"tags": tags}

    if action == "remove":
        if not name:
            raise ValueError("name requis.")
        if name not in tags:
            raise ValueError(f"Tag '{name}' introuvable.")
        del tags[name]
        _save(tags)
        return {"tags": tags}

    if action == "send":
        if not name:
            raise ValueError("name requis.")
        if name not in tags:
            raise ValueError(f"Tag '{name}' introuvable.")
        if not channel_id:
            raise ValueError("channelId requis pour envoyer un tag.")
        channel = await fetch_channel(client, channel_id)
        if not channel:
            raise ValueError(f"Salon {channel_id} introuvable.")
        await channel.send(tags[name])
        return {"sent": True, "name": name, "channelId": channel_id}

    raise ValueError(f"Action tag inconnue : '{action}'")


async def callback(client, message, args):
    if not args:
        return await message_edit(message, "`❌` **Usage :** `.tag <nom>`")
    name = args[0]
    tags = _load()
    if name not in tags:
        return await message_edit(message, f"`❌` **Tag `{name}` introuvable.**")
    return await message_edit(message, tags[name])
